libevmasm/Assembly.cpp

```diff
@@ -297,6 +503,142 @@ Json::Value Assembly::assemblyJSON(std::map<std::string, unsigned> const& _sourc
+		for (Json::ValueConstIterator dataIter = data.begin(); dataIter != data.end(); dataIter++)
+		{
+			solAssert(dataIter.key().isString());
+			std::string dataItemID = dataIter.key().asString();
+			Json::Value const& dataItem = data[dataItemID];
+			if (dataItem.isString())
+			{
+				solRequire(
+					dataItem.asString().empty() || !fromHex(dataItem.asString()).empty(),
+					AssemblyImportException,
+					"The value for key '" + dataItemID + "' inside '.data' is not a valid hexadecimal string."
+				);
+				result->m_data[h256(fromHex(dataItemID))] = fromHex(dataItem.asString());
+			}
+			else if (dataItem.isObject())
+			{
+				size_t index{};
+				try
+				{
+					// Using signed variant because stoul() still accepts negative numbers and
+					// just lets them wrap around.
+					int parsedDataItemID = std::stoi(dataItemID);
+					solRequire(parsedDataItemID >= 0, AssemblyImportException, "The key '" + dataItemID + "' inside '.data' is out of the supported integer range.");
+					index = static_cast<size_t>(parsedDataItemID);
+				}
+				catch (std::invalid_argument const&)
+				{
+					solThrow(AssemblyImportException, "The key '" + dataItemID + "' inside '.data' is not an integer.");
+				}
+				catch (std::out_of_range const&)
+				{
+					solThrow(AssemblyImportException, "The key '" + dataItemID + "' inside '.data' is out of the supported integer range.");
+				}
+
+				if (result->m_subs.size() <= index)
+					result->m_subs.resize(index + 1);
+				auto [subassembly, emptySourceList] = Assembly::fromJSON(dataItem, _level == 0 ? parsedSourceList : _sourceList, _level + 1);
+				solAssert(subassembly);
+				solAssert(emptySourceList.empty());
+				result->m_subs[index] = subassembly;
+			}
+			else
+				solThrow(AssemblyImportException, "The value of key '" + dataItemID + "' inside '.data' is neither a hex string nor an object.");
+		}
+	}
+
+	if (_level == 0)
+		result->encodeAllPossibleSubPathsInAssemblyTree();
+
+	return std::make_pair(result, _level == 0 ? parsedSourceList : std::vector<std::string>{});
+}
+
+void Assembly::encodeAllPossibleSubPathsInAssemblyTree(std::vector<size_t> _pathFromRoot, std::vector<Assembly*> _assembliesOnPath)
+{
+	_assembliesOnPath.push_back(this);
+	for (_pathFromRoot.push_back(0); _pathFromRoot.back() < m_subs.size(); ++_pathFromRoot.back())
+	{
+		for (size_t distanceFromRoot = 0; distanceFromRoot < _assembliesOnPath.size(); ++distanceFromRoot)
+			_assembliesOnPath[distanceFromRoot]->encodeSubPath(
+				_pathFromRoot | ranges::views::drop_exactly(distanceFromRoot) | ranges::to<std::vector>
+			);
+
+		m_subs[_pathFromRoot.back()]->encodeAllPossibleSubPathsInAssemblyTree(_pathFromRoot, _assembliesOnPath);
+	}
+}
+
+std::shared_ptr<std::string const> Assembly::sharedSourceName(std::string const& _name) const
+{
+	if (s_sharedSourceNames.find(_name) == s_sharedSourceNames.end())
+		s_sharedSourceNames[_name] = std::make_shared<std::string>(_name);
+
+	return s_sharedSourceNames[_name];
```

Comment:
```suggestion
	auto [it, _] = s_sharedSourceNames.try_emplace(_name, std::make_shared<std::string>(_name));

	return *it;
```

Reply:
This will unnecessarily allocate a new string even if not needed. The string is expected to be reused in most cases, so I think the original code is better here.

Reply:
Allocate a new string where? As in `std::make_shared<std::string>(_name)`?

Reply:
Yes.